README.md
bismark_v0.8.1
==============
Author: Peter Hickey (peter.hickey@gmail.com)

Branch of `bismark (v0.8.1)` to implement some of my ideas regarding `FLAG` values in `SAM/BAM` files. More specifically, to implement a proposed patch regarding how some of the strand and read orientation information is encoded in `SAM/BAM` files for paired-end data produced by Bismark and ensuring that other Bismark tools (such as `bismark_methylation_extractor`) can process these new file.

My motivation is for the suggested changes is because `SAM/BAMs` produced by Bismark don't always play nice with downstream tools for a few reasons:

(1) __FLAG in SAM/BAMs produced by Bismark mean paired-end read orientation is not compliant with SAM specifications__
GATK cannot correctly display paired-end data owing to the unique way Bismark encodes the bitwise FLAG. I've attached a screenshot from IGV of data PE data mapped using the Bismark defaults versus using the proposed patch. This is a problem for other downstream tools that rely on FLAGs for inferring the orientation of paired-end reads. This is implemented in the patched version via the --new_flag option. 
I know your FLAG encoding was a deliberate design choice because of the extra complications with stranded-ness for BS-seq data. However, even if Bismark uses these flags internally I think it should be possible to produce a SAM/BAM file that complies with SAM specifications.

(2) __Appending of /1 and /2 to paired-end readnames__
Picard's MarkDuplicates tool won't work by default when readnames are appended by /1 and /2 since they rely (in their default mode) on exact matching of readnames (it also won't work because of the FLAG issue). Similar problems occur when using some other Picard tools, such as when sorting SAM/BAM by queryname rather than coordinate using Picard's SortSam tool. Since the read_1/read_2 information can be encoded in the FLAG it shouldn't be necessary to also include it in the readnames. The fix is hardcoded in the patched version but could obviously be a command line option.

comethylation_v3.py is from https://github.com/PeteHaitch/Co-methylation based on commit 62f45088c57d75b23c152d3527db7ceeb8054384 (private repository)


RELEASE_NOTES.txt
RELEASE NOTES FOR Bismark v0.8.1 (16 Jul 2013)
-----------------------------------------------

Bismark
=======

Changed the way how the alignment overview file is being named to not generate a warning
message.


Methylation Extractor
=====================

Changed the function of '--ignore <int>' to ignore the first <int> bp from the 5' end of only
single-end reads or Read 1 of paired-end files. In addition, added a new option 
'--ignore_r2 <int>' to ignore the first <int> bp from the 5' end of Read 2 of paired-end files.
Since the first couple of bases in Read 2 of BS-Seq experiments show a severe bias towards
non-methylation as a result of the end-repair of sonicated fragments with unmethylated cytosines
(see M-bias plot), it is recommended that the the first couple of bp of Read 2 are removed before
starting downstream analysis. Please see the section on M-bias plots in the Bismark User Guide
for more details.

Changed colours, legends and background colour of the M-bias plot, but I think Simon 
is still not happy...





RELEASE NOTES FOR Bismark v0.8.0 (12 Jul 2013)
-----------------------------------------------

Bismark
=======

Added new option '--prefix <prefix>' to add <prefix> to the output filenames. Trailing
dots will be replaced by a single one. For example. '--prefix test' with 'file.fq'
would result in the output file 'test.file.fq_bismark.sam' etc.

Fixed a warning message that occurred when chromosomal sequences could not be extracted
in paired-end Bowtie2 mode.

Bismark will now generate a pie chart with the alignment statistics when a run has finished;
this allows to get a quick overview of how many sequences aligned uniquely, sequences that
did not align, either due to producing no alignment at all, multiple mapping or because it
was impossible to extract the chromosomal sequence. Drawing this plot will require the Perl
module GD::Graph; if it is not found on the system the plot will be skipped.


Methylation Extractor
=====================

Upon completion, the methylation extractor will now produce an M-bias (methylation bias)
plot, which shows the methylation proportion across each possible read position (described
in:  Hansen et al., Genome Biology, 2012, 13:R83). The data for the M-bias plot will be
written into a text file (to generate graphs by alternative means) and drawn into a .png
plot which  requires the Perl module GD::Graph; if GD::Graph cannot be found on the system,
only the table will be printed. The plot also contains the absolute number of methylation
calls per position.





RELEASE NOTES FOR Bismark v0.7.12 (10 May 2013)
-----------------------------------------------

Bismark
=======

Removed a rogue sleep(1) command that would slow down single-end Bowtie 2 alignments
for a single lane of HiSeq (200M sequences) from ~1 day to 6 years and 4 months (roughly).


bismark2bedGraph
================

bismark2bedGraph now keeps track of the temp files it just created instead of using
all files in the output folder ending in ".methXtractor.temp". This lets you kick off
the bedGraph conversion step from already sorted, individual methXtractor.temp files if
desired.




RELEASE NOTES FOR Bismark v0.7.11 (22 Apr 2013)
----------------------------------------------

Bismark
=======

Fixed non-functional single-end alignments with Bowtie2 which were accidentally broken
by introducing the option '--pbat' in v0.7.10 (an evil 'if' instead of 'elsif'...).

For paired-end alignments with Bowtie 1, the option '--non_bs_mm' would accidentally
confuse the number of mismatches of read 1 and read 2 whenever the first read aligned
in reverse orientation, i.e. for OB and CTOT alignments. This has now been corrected.
 
Previously, the option '--non_bs_mm' would potentially output non-integer values for
Bowtie 2 alignments if the read (or reference) contained 'N' characters. Alignment 
scores from 'N's are now adjusted so that they count as mismatches similar to what
Bowtie 1 does. This works for fine reads with up to and including 5 N's (which is quite
a lot...). 


Methylation Extractor
=====================

To avoid duplication and keep code modular, the bedGraph conversion step invoked by
the option '--bedGraph' is now been farmed out to the module 'bismark2bedGraph'. This
script is independent of the methylation extractor and also works as a stand-alone tool
from the methylation extractor output (compressed or gzip compressed files). To work
well from within the methylation extractor this script (which is now included in the
Bismark package) needs to reside in the same folder as the 'bismark_methylation_extractor'
itself.

bismark2bedGraph

Temporary chromosome files now have an input file name included in their file name to 
enable parallel processing of several files in the same directory at the same time.


To avoid duplication and keep code modular, the bedGraph to genome-wide cytosine methylation
report step invoked by the option '--cytosine_report' has now been split out to the
module 'bedGraph2cytosine'. This script is independent of the methylation extractor and
also works as a stand-alone tool from the Bismark bedGraph '--counts' output (compressed
or gzip compressed files). To work well from within the methylation extractor this 
script (which is now included in the Bismark package) needs to reside in the same folder
as the 'bismark_methylation_extractor' itself.


Deduplication script
====================

Fixed some warnings that were thrown if '--bam' was not specified.



RELEASE NOTES FOR Bismark v0.7.10 (18 Apr 2013)
----------------------------------------------

Bismark
=======

Added new option '--gzip' that causes temporary bisulfite conversion files to be
written out in a GZIP compressed form in order to save disk space. This option is
available for most alignment modes but is not available for paired-end FastA files 
(not many people use PE FastA files...). This option might be somewhat slower than

testing_new_flag.sh
#### DESCRIPTION ####
# Workflow for testing my modifications to bismark (v0.8.1_PH). 
# To be run in folder containing my modified version of bismark (v0.8.1_PH)
# Peter Hickey (peter.hickey@gmail.com)
# 25/07/2013

#### Download SRA data SRR400564 and extract as gzipped FASTQ files ####
mkdir old_flag
mkdir new_flag
mkdir SRR400564

cd SRR400564

wget http://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR400/SRR400564/SRR400564.sra 

fastq-dump --gzip --split-files SRR400564.sra

#### Align data both with and without the --new_flag option ####
cd ..

./bismark --bam --new_flag --prefix new_flag /home/users/lab0605/hickey/WGBS/aligner_indexes/Bismark_Bowtie1/hg19+lambda_phage/Bismark_Bowtie1_hg19+lambda_phage.genome_folder/ -1 SRR400564/SRR400564_1.fastq.gz -2 SRR400564/SRR400564_2.fastq.gz -o new_flag/

./bismark --bam --prefix old_flag /home/users/lab0605/hickey/WGBS/aligner_indexes/Bismark_Bowtie1/hg19+lambda_phage/Bismark_Bowtie1_hg19+lambda_phage.genome_folder/ -1 SRR400564/SRR400564_1.fastq.gz -2 SRR400564/SRR400564_2.fastq.gz -o old_flag/

#### Run bismark_methylation_extractor on new_flag.SRR400564_1.fastq.gz_bismark_pe.bam and old_flag.SRR400564_1.fastq.gz_bismark_pe.bam ####
./bismark_methylation_extractor -p --no_overlap --ignore 2 --ignore_r2 7 -o old_flag --bedGraph --counts old_flag/old_flag.SRR400564_1.fastq.gz_bismark_pe.bam

./bismark_methylation_extractor -p --no_overlap --ignore 2 --ignore_r2 7 -o new_flag --bedGraph --counts new_flag/new_flag.SRR400564_1.fastq.gz_bismark_pe.bam

#### Check for differences between new_flag and old_flag files, e.g. using diff ####



test_data/explore_methylation_calling.r
#### DESCRIPTION ####
# Explore the differences in the methylation calls between bismark (v0.8.1) with --new_flag option and without --new_flag option
# Peter Hickey (peter.hickey@gmail.com)
# 24/07/2013

##### Read in data ####
x <- list(new_flag_XM_filter = read.table('new_flag_filter=XM.wf', header = T, sep = '\t', stringsAsFactors = FALSE),
          new_flag_sequence_filter = read.table('new_flag_filter=sequence.wf', header = T, sep = '\t', stringsAsFactors = FALSE),
          new_flag_bismark_filter = read.table('new_flag_filter=bismark.wf', header = T, sep = '\t', stringsAsFactors = FALSE), 
          new_flag_no_filter = read.table('new_flag_filter=none.wf', header = T, sep = '\t', stringsAsFactors = FALSE),
          new_flag_bismark_methylation_extractor = read.table('CS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam.bedGraph', header = F, stringsAsFactors = FALSE, sep = '\t'),
          old_flag_bismark_methylation_extractor = read.table('CS_old_flag.SRR400564_1.fastq.gz_bismark_pe.bam.bedGraph', header = F, stringsAsFactors = FALSE, sep = '\t')
)

#### Dimension of the different methylation call sets ####
sapply(X = x, FUN = dim)

#### Comparison between new_flag_bismark_filter and old_flag_bismark_methylation_extractor ####
## Positions unique to one or the other
unique_positions <-  

testing.sh
#### DESCRIPTION ####
# Workflow for testing my modifications to bismark (v0.8.1_PH). To be run in folder containing my modified version of bismark (v0.8.1_PH)
# Peter Hickey (peter.hickey@gmail.com)
# 24/07/2013

#### Environment variables ####
TMP_DIR=/usr/local/work/hickey/

#### Download SRA data and extract as gzipped FASTQ files ####
mkdir test_data

cd test_data

wget http://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR400/SRR400564/SRR400564.sra

fastq-dump --gzip --split-files SRR400564.sra

#### Align data both with and without the --new_flag option ####
cd ..

./bismark --bam --new_flag --prefix new_flag /home/users/lab0605/hickey/WGBS/aligner_indexes/Bismark_Bowtie1/hg19+lambda_phage/Bismark_Bowtie1_hg19+lambda_phage.genome_folder/ -1 test_data/SRR400564_1.fastq.gz -2 test_data/SRR400564_2.fastq.gz -o test_data/

./bismark --bam --prefix old_flag /home/users/lab0605/hickey/WGBS/aligner_indexes/Bismark_Bowtie1/hg19+lambda_phage/Bismark_Bowtie1_hg19+lambda_phage.genome_folder/ -1 test_data/SRR400564_1.fastq.gz -2 test_data/SRR400564_2.fastq.gz -o test_data/

#### Sort BAM files by coordinate and index the sorted BAMs ####
cd test_data

SortSam I=new_flag.SRR400564_1.fastq.gz_bismark_pe.bam O=CS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam SO=coordinate TMP_DIR=${TMP_DIR}

SortSam I=old_flag.SRR400564_1.fastq.gz_bismark_pe.bam O=CS_old_flag.SRR400564_1.fastq.gz_bismark_pe.bam SO=coordinate TMP_DIR=${TMP_DIR}

samtools index CS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam

samtools index CS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam

#### Sort BAMs by queryname ####
SortSam I=new_flag.SRR400564_1.fastq.gz_bismark_pe.bam O=QS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam SO=queryname TMP_DIR=${TMP_DIR}

SortSam I=old_flag.SRR400564_1.fastq.gz_bismark_pe.bam O=QS_old_flag.SRR400564_1.fastq.gz_bismark_pe.bam SO=queryname TMP_DIR=${TMP_DIR}

#### Run comethylation_v3.py on QS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam with different filters on overlapping mates from paired-end reads ####
mkdir test_data/log

python comethylation_v3.py --nTuple 1 --methylationType CG --overlappingPairedEndCheck bismark test_data/QS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam test_data/new_flag_filter=bismark &> test_data/log/new_flag_filter=bismark.log

python comethylation_v3.py --nTuple 1 --methylationType CG --overlappingPairedEndCheck sequence test_data/QS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam test_data/new_flag_filter=sequence &> test_data/log/new_flag_filter=sequence.log

python comethylation_v3.py --nTuple 1 --methylationType CG --overlappingPairedEndCheck XM test_data/QS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam test_data/new_flag_filter=XM &> test_data/log/new_flag_filter=XM.log

python comethylation_v3.py --nTuple 1 --methylationType CG --overlappingPairedEndCheck none test_data/QS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam test_data/new_flag_filter=none &> test_data/log/new_flag_filter=none.log

#### Run bismark_methylation_extractor and bismark2bedGraph with --comprehensive flag on CS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam and CS_old_flag.SRR400564_1.fastq.gz_bismark_pe.bam. Run it on both the coordianate sorted BAMs (CS_*.bam) and queryname sorted BAMs (QS_*.bam) to see if there is any difference  ####
./bismark_methylation_extractor -p --no_overlap --comprehensive --ignore 0 --ignore_r2 0 --merge_non_CpG -o test_data --bedGraph --counts --buffer_size 10G test_data/CS_old_flag.SRR400564_1.fastq.gz_bismark_pe.bam

./bismark_methylation_extractor -p --no_overlap --comprehensive --ignore 0 --ignore_r2 0 --merge_non_CpG -o test_data --bedGraph --counts --buffer_size 10G test_data/CS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam

./bismark_methylation_extractor -p --no_overlap --comprehensive --ignore 0 --ignore_r2 0 --merge_non_CpG -o test_data --bedGraph --counts --buffer_size 10G test_data/QS_old_flag.SRR400564_1.fastq.gz_bismark_pe.bam

./bismark_methylation_extractor -p --no_overlap --comprehensive --ignore 0 --ignore_r2 0 --merge_non_CpG -o test_data --bedGraph --counts --buffer_size 10G test_data/QS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam

#### Explore results in R (see test_data/explore_methylation_calling.r) ####

#### Some timing comparisons (comethylation_v3.py is faster but at the expense of memory consumption ####
mkdir test_data/timing
time python comethylation_v3.py --nTuple 1 --methylationType CG --overlappingPairedEndCheck bismark test_data/QS_new_flag.SRR400564_1.fastq.gz_bismark_pe.bam test_data/timing/new_flag_filter=bismark
#real    16m49.769s
#user    16m30.252s
#sys     0m11.577s

time ./bismark_methylation_extractor -p --no_overlap --comprehensive --ignore 0 --ignore_r2 0 --merge_non_CpG--bedGraph --counts --buffer_size 10G test_data/QS_old_flag.SRR400564_1.fastq.gz_bismark_pe.bam
#real    95m32.997s
#user    96m48.782s
#sys     1m21.358s
